src/toplevelprops.ts

```diff
@@ -20,15 +21,49 @@ export interface TopLevelProperties {
   padding?: Padding;
 
   /**
-   * Resize is a boolean indicating if autosize layout should be re-calculated on every update.
+   * Sets how the visualization size should be determined. If a string, should be one of `"pad"`, `"fit"` or `"none"`.
+   * Object values can additionally specify parameters for content sizing and automatic resizing.
+   *
+   * __Default value__: `pad`
+   */
+  autosize?: AutosizeType | Autosize;
+}
+
+export type AutosizeType = 'pad' | 'fit' | 'none';
+
+export interface Autosize {
+  /**
+   * The sizing format type. One of `"pad"`, `"fit"` or `"none"`. See the [autosize type](https://vega.github.io/vega/docs/specification/#autosize) documentation for descriptions of each.
+   *
+   * __Default value__: `"pad"`
+   */
+  type?: AutosizeType;
+
+  /**
+   * A boolean flag indicating if autosize layout should be re-calculated on every view update.
    *
    * __Default value__: `false`
    */
-  autoResize?: boolean;
+  resize?: boolean;
+
+  /**
+   * Determines how size calculation should be performed, one of `"content"` or `"padding"`. The default setting (`"content"`) inteprets the width and height settings as the data rectangle (plotting) dimensions, to which padding is then added. In contrast, the `"padding"` setting includes the padding within the view size calculations, such that the width and height settings indicate the **total** intended size of the view.
+   *
+   * __Default value__: `"content"`
+   */
+  contains?: 'content' | 'padding';
+}
+
+export function isAutosize(a: Autosize | AutosizeType): a is Autosize {
+  return !isString(a);
+}
+
+export function autoSizeFit(autosize: AutosizeType | Autosize) {
```

Comment:
See this [comment](https://github.com/vega/vega-lite/pull/2991#discussion_r137980261)

Reply:
I don't like `(inputSpec.autosize as AutoSizeParams).type === 'fit'`

Reply:
Fixed differently